[5기 - 이중원] 1~2주차 과제: 도서 관리 어플리케이션 제출합니다.

shoeone96 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

![Flowchart](https://github.com/prgrms-be-devcourse/java-library-management/assets/85065626/660470b4-7794-49f8-8b71-2a7657beaf1a)
다음의 흐름으로 진행되는 도서관 시스템 구축

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->

- [x] 도서등록
- [x] 전체 도서 목록 조회
- [x] 제목으로 도서 검색
- [x] 도서 대여
- [x] 도서 반납
- [x] 도서 분실
- [x] 도서 삭제
- [x] 각 상황별 예외처리
- [x] 각 상황별 테스트 케이스 작성(controller 부분은 추가 테스트 필요)  

structure
- controller
    - **BookController**(메서드에 따라 필요한 정보를 Console로 입출력하는 클래스)
    - **NumberMappingHandler**(원하는 요청 옵션을 해당 Controller method에 매핑해주는 클래스)
- domain
    - **Book**(책 도메인, 책 상태 관련 메서드 포함)
    - **BookStatus**(책 상태를 나타내는 enum)
- exception
    - **ErrorChecking**(예외를 점검하는 메서드를 모아 놓은 클래스)
    - **ErrorCode**(예외 내용을 정리해놓은 enum)
    - **LibraryException**(메시지 전달 등을 위한 custom Exception)
- front
    - **BookConsole** (입출력 관련 메서드를 모은 클래스)
- repository
    - **BookRepository** (db Repository interface)
    - **CSVBookRepository** (CSV 파일을 db로 사용하는 클래스)
    - **TestBookRepository** (in-memory list 를 db로 사용하는 클래스)
- service
    - **BookService** (Controller에서 받아온 정보를 토대로 로직을 실행하는 서비스 클래스)
- **BookAplication(실행 메서드)**
- **LibraryConfiguration(의존성 주입 클래스)**
- **LibraryEngine(서비스 구동 클래스)**



## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
### 1. 현재 코드가 객체지향적으로 짜여졌는지 궁금합니다.  
  - 객체지향적으로 코드를 짜려고 하니 스프링 구조를 많이 떠올려 참고를 했습니다.  

### 2. 이번 과제에서 메서드 단위로 커밋을 안하고 코드 작성 후 커밋을 진행하다보니 커밋 단위가 평소보다 컸습니다.  
  - 원래는 메서드 단위로 커밋을 하는 편인데 이상적인 커밋 단위는 어떻게 잡는 게 좋을까요?  

### 3. 예외처리 부분을 어떻게 더 효과적으로 처리할 수 있을지 궁금합니다.  
  - 기존에는 스프링 위주로 프로젝트를 진행했었습니다.
  - 그래서 @ControllerAdvice를 이용한 전역처리 위주로 진행했었는데 직접 try-catch 및 재실행을 하는 부분을 구현하다보니 코드가 지저분해 보여서 더 객체지향적으로 잘 짤 수 있는 방법이 궁금합니다.  

### 4. CSV 파일을 구현한 부분에서 어댑터 패턴을 이용해 조금 더 코드를 분리하는 게 좋은지 궁금합니다.  

### 5. 테스트 코드는 어떤 단위로 만들고 얼마나 만들면 좋을지 궁금합니다.
  - 현재 controller 부분은 입출력 관련 테스트가 제대로 작동하지 않아 수정 중에 있습니다.   

**무엇보다 지금까지 코드에 대해 피드백을 받아본 적이 거의 없어 전체적으로 제 문제점이 뭔지 메타인지가 잘 안되는 문제가 있습니다. 가감없이 말씀 부탁드릴게요!**

shoeone96 committed:
Update README.md

shoeone96 committed:
[chore] gitignore 세팅

shoeone96 committed:
[chore] 초기 세팅

shoeone96 committed:
[feat] library interface 작성

shoeone96 committed:
[feat] library Interface에 맞춰 인메모리 db 작성

shoeone96 committed:
[feat] library interface 맞춰 CSV 파일 읽는 repository 작성

shoeone96 committed:
[feat] CSV CURD를 위한 파일 생성

shoeone96 committed:
예외 처리를 위한 Error 내용 및 Exception 클래스 생성

shoeone96 committed:
[feat] Library에서 사용되는 책 도메인 생성

shoeone96 committed:
[feat] Console 출력을 위한 클래스 작성

shoeone96 committed:
[feat] service 계층 코드 작성

shoeone96 committed:
[feat] service 코드를 받아올 컨트롤러 코드 작성

shoeone96 committed:
[feat] 사용자 입력과 매칭시켜주는 mapping 클래스 작성

shoeone96 committed:
[feat] 사용자 작성에 다라 의존성을 주입해주는 configuration 코드 작성

shoeone96 committed:
Library 작동 및 의존성 세팅을 위한 클래스 작성

shoeone96 committed:
[feat] 실제 애플리케이션 작동 코드

shoeone96 committed:
[Test] 각 계층별 테스트 코드 작성

shoeone96 committed:
[chore] 기타 세팅

shoeone96 committed:
Merge remote-tracking branch 'origin/main' into main

shoeone96 committed:
[test] Test 코드 작성(controller, service, repository)

아직 controller test에 대하여 입출력 관련 오류가 발생하여 추가적으로 테스트 수정중
-> 입력 단어를 넣어도 제대로 입력되지 않는 문제

shoeone96 committed:
[feat] repository 및 service로직에 db 내용을 비울 수 있는 메서드 추가

- 테스트 코드 작성 등을 위해 db를 비울 수 있는 메서드 추가

SeokRae commented on src/test/java/com/programmers/repository/TestBookRepositoryTest.java:
```diff
@@ -0,0 +1,93 @@
+        bookList = bookRepository.findAll();
+
+        //then
+        Assertions.assertThat(bookList.size()).isEqualTo(4);
```
```suggestion
        assertThat(bookList).hasSize(4);
```

사이즈를 검증하는 assertJ 표현법을 활용해보시죠

shoeone96 replied:
[[refactor] 1. 테스트 코드 컬렉션 사이즈 비교 메서드 변경](https://github.com/prgrms-be-devcourse/java-library-management/commit/163babd87385dc4619bd8222ca36ca3ab08af260)

반영 완료했습니다!

SeokRae commented on src/test/java/com/programmers/repository/TestBookRepositoryTest.java:
```diff
@@ -0,0 +1,93 @@
+
+        //then
+        Assertions.assertThat(bookList.size()).isEqualTo(4);
+        Assertions.assertThat(book).isEqualTo(bookList.get(bookList.size()-1));
```
book의 toString 값 출력 값 전체를 검증한다는 것에 대해서는 이해를 했지만,
보다 구체적인 방법으로 확인하는 것이 필요해보입니다.

shoeone96 replied:
[[refactor] 2. 테스트 코드 데이터 비교 테스트 개선](https://github.com/prgrms-be-devcourse/java-library-management/commit/d44e9aa6dbc6dbd2397378ec3bf22a00c665fc18)

반영 완료 했습니다!

SeokRae commented on src/main/java/com/programmers/exception/ErrorChecking.java:
```diff
@@ -0,0 +1,26 @@
+package com.programmers.exception;
+
+public class ErrorChecking {
+
```
```suggestion
    private ErrorChecking() {}
```

유틸 클래스를 사용하는 경우 private 생성자를 통해 불필요한 객체가 생성되지 않도록 하는 것이 좋습니다.!

shoeone96 replied:
[[refactor] 3. 유틸 클래스 기본 생성자의 접근제어자 변경
](https://github.com/prgrms-be-devcourse/java-library-management/commit/9fdf074903999260c9e3348a9d805558b19e441d)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/domain/Book.java:
```diff
@@ -0,0 +1,102 @@
+
+public class Book {
+    @Override
+    public String toString() {
```
toString은 어쩐일로 이렇게 올라와 있는 걸까요? 
equals hashCode / toString 은 일반적으로 최하단에 위치하게 됩니다.

클래스를 봤을 때, 해당 클래스가 어떠한 역할을 한다 라는 의미를 갖는 필드 / 생성자 / 메서드(public, private) 등이 나열되게 됩니다.

shoeone96 replied:
[[refactor] 4. 코드 가독성을 위해 메서드 위치 변경
](https://github.com/prgrms-be-devcourse/java-library-management/commit/fe06651a576084cf34ba8a236cfe3cad7b89ad96)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/domain/Book.java:
```diff
@@ -0,0 +1,102 @@
+    private BookStatus bookStatus;
+
+    public Book(String title, String author, int totalPageNumber) {
+        this.title = title;
+        this.author = author;
+        this.totalPageNumber = totalPageNumber;
+        this.bookStatus = BookStatus.AVAILABLE;
```
```suggestion
        this(null, title, author, totalPageNumber, BookStatus.AVAILABLE);
```

this()를 활용하여 생성자를 통해 초기화용 생성자와 오버로딩 생성자를 구분해서 사용하면 초기화 시점에 대해서 명확하게 인지할 수 있습니다.

shoeone96 replied:
[[refactor] 5. 생성자 작성 방식 변경](https://github.com/prgrms-be-devcourse/java-library-management/commit/c8fe762d81fcaffc24d0fc68cc437afa36649891)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/domain/Book.java:
```diff
@@ -0,0 +1,102 @@
+        this.bookStatus = bookStatus;
+    }
+
+    public static Book enrollingBook(String title, String author, int totalPageNumber){
```
정적 팩토리 메서드 패턴을 사용하여 메서드에 생성의 의미를 갖는 메서드를 제공하기 위해서는 위의 생성자의 접근 제한자를 막도록 하여, 진입로를 제한하는 것을 추천드립니다.

shoeone96 replied:
[[refactor] 6. 정적 팩토리 메서드에 사용되는 생성자의 접근제어자 변경](https://github.com/prgrms-be-devcourse/java-library-management/commit/ae19006f7d52c5bd8e9a05de42dd1e6d2718f456)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/front/BookConsole.java:
```diff
@@ -0,0 +1,170 @@
+        System.out.println();
+    }
+
+    public static String inputBookName(){
```
비슷한 패턴을 갖는 출력문이라면 파라미터 하나 추가해서 공통으로 사용할 수도 있을 것 같습니다.

shoeone96 replied:
[[refactor] 7. Console 유틸 클래스에서 공통 부분 추출 가능한 메서드 리팩토링
](https://github.com/prgrms-be-devcourse/java-library-management/commit/5dd7788b55357c1778e29690e93839769d78ec0a)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/repository/TestBookRepository.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public Optional<Book> findByBookId(Long bookId) {
+        return bookList.stream()
+                .filter((book) -> Objects.equals(book.getBookId(), bookId))
```
객체지향 적인 코드를 시도하는 것이니 bookId 값도 비교하는 메서드를 Book 클래스 내부로 숨기는 시도를 해봐도 좋을 듯 합니다.

shoeone96 replied:
[[refactor] 8. id로 책을 조회하는 메서드 Book class 내부에 구현](https://github.com/prgrms-be-devcourse/java-library-management/commit/7919c1cf1bc25e49b9229ea72ff3adaba2b1bb1b)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/domain/Book.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private Long bookId;
+    private String title;
+    private String author;
+    private int totalPageNumber;
```
```suggestion
    private final String title;
    private final String author;
    private final int totalPageNumber;
```

크게 변경사항이 없는 필드로 보여집니다. 이러한 필드는 final 로 선언하여 최대한 불변 객체가 만들어질 수 있도록 하는 것이 좋습니다.

shoeone96 replied:
[[refactor] 9. 변경사항이 없는 필드 final 제어자 추가](https://github.com/prgrms-be-devcourse/java-library-management/commit/637b1755aa06b015ba4b8734184c189163a8858a)

반영 완료했습니다!

SeokRae commented on src/main/java/com/programmers/repository/CSVBookRepository.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Override
+    public void saveBook(Book saveBook) {
+        saveBook.settingId(++sequence);
```
setter를 제공해서 id를 변경하는 것도 좋지만, 차라리 Book 클래스에 book과 sequence를 파라미터로하는 생성자 또는 정적 팩토리 메서드를 제공하여 새로운 객체를 만들고 불변 객체를 지향하는 방향이 더 좋을 것 같습니다.

shoeone96 replied:
[[refactor] 10. Book setter 메서드(settingId()) 제거](https://github.com/prgrms-be-devcourse/java-library-management/commit/5fbb47623513ceacd769313614372ca6557d0d66)

SeokRae commented on src/main/java/com/programmers/controller/NumberMappingHandler.java:
```diff
@@ -0,0 +1,45 @@
+        this.bookController = bookController;
+    }
+
+    public boolean numberMapping(String num){
```
메서드의 라인 수를 줄일 수 있는 방법을 고민해보시는 것을 추천드립니다.
수정 사항이 생기는 경우 해당 메서드를 분석해야 하는 시간이 단축되는 효과를 볼 수 있습니다.

shoeone96 replied:
[[refactor] 11. 메서드 분리 실행](https://github.com/prgrms-be-devcourse/java-library-management/commit/f147e8f3bc8e870d4ea03633b7182ab5d3b18d4b)

SeokRae reviewed (approved):
안녕하세요. 중원님 멘토 알입니다.

첫 번째 과제 진행하시느라 고생 많이 하셨습니다.

코드 리뷰하면서 고민해볼 만한 내용에 대해서 작성하였습니다.

### ✅ PR 포인트 & 궁금한 점

1. 현재 코드가 객체지향적으로 짜여졌는지 궁금합니다.

스프링 프레임워크는 수년 간의 하위 호환성 및 개선 등이 이루어져있기 때문에 다양한 패턴이 얽혀있다고 생각해야 할 것 같습니다. 현재 과제에서는 자바 문법으로 구현할 수 있는 방법을 생각해야 할 것 같습니다.
그 기준의 시작을 SOLID로 잡고, 구현한 내용을 그에 대입하여 설명할 수 있다면 객체지향에 대해서 논의할 수 있을 것 같습니다.

2. 이번 과제에서 메서드 단위로 커밋을 안하고 코드 작성 후 커밋을 진행하다보니 커밋 단위가 평소보다 컸습니다.

과제 이기 때문에 한번에 올릴 수 밖에 없는 상황임을 인지하고 있습니다.
회사마다 커밋 단위의 기준이 다르고 기능 개발의 라인 수와 수정 개발의 라인 수 차이가 다르기 때문에 정답인 기준은 없을 것 같습니다.
그래도 굳이 기준을 잡자면 개인적으로는 기능 하나당 2 ~ 30 라인이면 되지 않을까 싶습니다.

3. 예외처리 부분을 어떻게 더 효과적으로 처리할 수 있을지 궁금합니다.

예외 처리를 객체지향적으로 처리 하는 방법이라는 것은 따로 없다고 생각됩니다.
예외가 발생할 수 있는 상황을 대처하는 방법으로는 try catch로 할 수도 있고, 방어코딩을 통해 진행할 수도 있습니다. 그리고 상황 발생 후 복원하는 과정을 어떻게 가져갈 것이냐에 대한 고민을 하다보면 부가적인 코드가 계속 추가되는 것은 불가피 합니다.
다만 이러한 부가적인 내용을 디자인 패턴을 통해 처리할 수도 있으나, 이러한 구조는 복잡도를 높이게 됩니다.최종적으로는 이러한 상황들이 모두 트레이드 오프이기 때문에 어느정도는 타협하는 과정이 필요합니다.

4. CSV 파일을 구현한 부분에서 어댑터 패턴을 이용해 조금 더 코드를 분리하는 게 좋은지 궁금합니다.

CSVBookRepository 클래스 내에 공통 로직이 있는지, 이를 분리할  있을지 에 대해서 먼저 고민해보시는 것을 추천합니다.
공통으로 분리할 수 있는 내용이 생기면 해당 메서드 들의 의미가 CSV 파일인지 Book 도메인 인지 구분이 될 것 같습니다.
그 이후 상황에서 리펙토링에 대해서 고민해볼 수 있을 것 같습니다.

5. 테스트 코드는 어떤 단위로 만들고 얼마나 만들면 좋을지 궁금합니다.

테스트 코드는 단위 테스트, 통합 테스트, 시나리오 테스트를 기준으로 작성해보시는 것을 추천드립니다.
클래스 하나, 메서드 하나의 로직을 검증하기 위해서는 어떤 의존성을 제외해야 하는지
통합 테스트는 어떤 규모로 작성해야 할 지, 프로젝트가 완성된 후에 최종적으로 어떤 흐름으로 검증할 수 있는지에 대한 시나리오 테스트 등이 있을 것 같습니다.

실제 업무에서는 이에 대한 기준을 내부에서 정하기 때문에 과제 및 개인 / 팀플에서는 다양한 시도를 해보시는 것을 권장합니다.

shoeone96 committed:
[refactor] 1. 테스트 코드 컬렉션 사이즈 비교 메서드 변경

변경 전: 사이즈 값을 구한 후 그 정수값을 테스트하는 메서드
변경 후: 리스트의 사이즈를 검증하는 메서드
E.g. assertThat(list.size()).isEqualTo();
-> assertThat(list).hasSize();

shoeone96 committed:
[refactor] 2. 테스트 코드 데이터 비교 테스트 개선

기존 두 데이터를 비교할 땡 toString()을 이용하여 한 데이터를 저장한 직후 해당 데이터를 호출해 값을 비교하는 방법 채용
-> 데이터를 받아온 객체에 equals 를 override하여 비교하는 방법으로 변경
Assertions.assertThat(savedBook).isEqualTo(findBook) 으로 비교하는 방식 변경

shoeone96 committed:
[refactor] 3. 유틸 클래스 기본 생성자의 접근제어자 변경

생성자를 이용하지 않는 클래스에 대하여 함부로 생성하지 못하도록 private 기본생성자로 변경

shoeone96 committed:
[refactor] 4. 코드 가독성을 위해 메서드 위치 변경

 equals(), toString(), hashCode() 등 본 역할이 아닌 Object의 override 메서드 위치를 가장 하단으로 변경

shoeone96 committed:
[refactor] 5. 생성자 작성 방식 변경

이미 AllArgsArgument 역할을 하는 생성자가 있는 상황에서
다른 생성자가 필요한 경우라 다른 생성자를 호출해 이 생성자의 의미를 더욱 명확하게 확인할 수 있게 표기

shoeone96 committed:
[refactor] 6. 정적 팩토리 메서드에 사용되는 생성자의 접근제어자 변경

정적 팩토리 메서드로 객체를 생성해서 반환하므로 여기에 사용되는 메서드가 외부에서 다른 용도로 사용되지 않도록 private 접근제어자로 변경

shoeone96 committed:
[refactor] 7. Console 유틸 클래스에서 공통 부분 추출 가능한 메서드 리팩토링

대여, 분실, 반납, 삭제에 필요한 출력 메서드를 하나의 메서드로 통합

shoeone96 committed:
[refactor] 8. id로 책을 조회하는 메서드 Book class 내부에 구현

객체 지향적 설계를 위해 Book class 내부에 직접 구현

shoeone96 committed:
[refactor] 9. 변경사항이 없는 필드 final 제어자 추가

변경해야 할 경우가 없어 변경 시 컴파일 오류가 나게 변경

shoeone96 committed:
[refactor] 10. Book setter 메서드(settingId()) 제거

1. 기존에 repository 단계에서 id 저장을 위해 사용하던 sequence를 전역 상수로 설정
2. 메서드를 통해서만 사용 및 값 변화가 가능하게 살정
3. 책을 등록하는 과정에서 상수 값을 사용해 bookId가 들어가게 변경
4. 책 Id가 생성자를 통해서만 설정될 수 있게 만듬
5. setter 메서드 제거
6. 각 구현한 repository 마다 존재하던 sequence 필드 제거

shoeone96 committed:
[refactor] 11. 메서드 분리 실행

각 기능 단위로 해당 클래스 내부에서만 사용가능한 메서드로 분리
